File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/GeneralDataService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;

import java.io.File;

import sk.styk.martin.apkanalyzer.model.detail.AppSource;
import sk.styk.martin.apkanalyzer.model.detail.GeneralData;
import sk.styk.martin.apkanalyzer.util.AndroidVersionHelper;
import sk.styk.martin.apkanalyzer.util.InstallLocationHelper;

/**
 * Created by Martin Styk on 30.06.2017.
 */
public class GeneralDataService {

    private PackageManager packageManager;

    public GeneralDataService(PackageManager packageManager) {
        this.packageManager = packageManager;
    }

    public GeneralData get(@NonNull PackageInfo packageInfo) {

        ApplicationInfo applicationInfo = packageInfo.applicationInfo;

        GeneralData generalData = new GeneralData();

        generalData.setPackageName(packageInfo.packageName);
        generalData.setVersionCode(packageInfo.versionCode);
        generalData.setVersionName(packageInfo.versionName);
        generalData.setInstallLocation(InstallLocationHelper.resolveInstallLocation(packageInfo.installLocation));
        generalData.setFirstInstallTime(packageInfo.firstInstallTime);
        generalData.setLastUpdateTime(packageInfo.lastUpdateTime);

        if (applicationInfo != null) {
            generalData.setIcon(applicationInfo.loadIcon(packageManager));
            CharSequence description = applicationInfo.loadDescription(packageManager);
            generalData.setDescription(description != null ? description.toString() : null);
            CharSequence label = applicationInfo.loadLabel(packageManager);
            generalData.setApplicationName(label != null ? label.toString() : applicationInfo.packageName);
            generalData.setProcessName(applicationInfo.processName);
            generalData.setSystemApp((applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0);
            generalData.setApkDirectory(applicationInfo.sourceDir);
            generalData.setApkSize(getApkSize(applicationInfo.sourceDir));
            generalData.setDataDirectory(applicationInfo.dataDir);
            generalData.setMinSdkVersion(AndroidManifestService.getMinSdkVersion(applicationInfo, packageManager));
            generalData.setMinSdkLabel(AndroidVersionHelper.resolveVersion(generalData.getMinSdkVersion()));
            generalData.setTargetSdkVersion(applicationInfo.targetSdkVersion);
            generalData.setTargetSdkLabel(AndroidVersionHelper.resolveVersion(applicationInfo.targetSdkVersion));
        }

        generalData.setSource(AppSource.get(packageManager, packageInfo.packageName, generalData.isSystemApp()));

        return generalData;
    }


    public long getApkSize(String sourceDir) {
        return new File(sourceDir).length();
    }

}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/GeneralDataService.kt
package sk.styk.martin.apkanalyzer.business.service

import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import sk.styk.martin.apkanalyzer.model.detail.AppSource
import sk.styk.martin.apkanalyzer.model.detail.GeneralData
import sk.styk.martin.apkanalyzer.util.AndroidVersionHelper
import sk.styk.martin.apkanalyzer.util.InstallLocationHelper
import java.io.File

/**
 * @author Martin Styk
 * @version 30.06.2017.
 */
class GeneralDataService {

    fun get(packageInfo: PackageInfo, packageManager: PackageManager): GeneralData {

        val applicationInfo = packageInfo.applicationInfo

        val generalData = GeneralData()

        generalData.packageName = packageInfo.packageName
        generalData.versionCode = packageInfo.versionCode
        generalData.versionName = packageInfo.versionName
        generalData.installLocation = InstallLocationHelper.resolveInstallLocation(packageInfo.installLocation)
        generalData.firstInstallTime = packageInfo.firstInstallTime
        generalData.lastUpdateTime = packageInfo.lastUpdateTime

        applicationInfo?.let {

            generalData.icon = applicationInfo.loadIcon(packageManager)
            generalData.description = applicationInfo.loadDescription(packageManager)?.toString()
            generalData.applicationName = applicationInfo.loadLabel(packageManager)?.toString() ?: applicationInfo.packageName
            generalData.processName = applicationInfo.processName
            generalData.isSystemApp = (applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM) != 0
            generalData.uid = applicationInfo.uid
            generalData.apkDirectory = applicationInfo.sourceDir
            generalData.apkSize = getApkSize(applicationInfo.sourceDir)
            generalData.dataDirectory = applicationInfo.dataDir
            generalData.minSdkVersion = AndroidManifestService.getMinSdkVersion(applicationInfo, packageManager)
            generalData.minSdkLabel = AndroidVersionHelper.resolveVersion(generalData.minSdkVersion)
            generalData.targetSdkVersion = applicationInfo.targetSdkVersion
            generalData.targetSdkLabel = AndroidVersionHelper.resolveVersion(applicationInfo.targetSdkVersion)

        }

        generalData.source = AppSource.get(packageManager, packageInfo.packageName, generalData.isSystemApp)

        return generalData
    }

    fun getApkSize(sourceDir: String): Long {
        return File(sourceDir).length()
    }

}
